packages/lit-dev-server/src/middleware/content-security-policy-middleware.ts

```diff
@@ -176,11 +183,19 @@ export const contentSecurityPolicyMiddleware = (
   const strictFallbackCsp = makePolicy(`default-src 'none'`);
 
   return async (ctx, next) => {
-    await next();
-
     let policy: string;
-    if (ctx.response.type === 'text/html') {
-      policy = htmlCsp;
+    // Note we can't rely on ctx.type being set by the downstream middleware,
+    // because for a 304 Not Modified response, the Content-Type header will not
+    // be set.
+    //
+    // Also note that we don't necessarily need to set a CSP on 304 responses at
+    // all, because the browser will use the one from the previous 200 response
+    // if missing (as it does for all headers). However, by including a CSP on
+    // 304 responses, we cover the case where the CSP has changed, but the
+    // file's content (and hence ETag) has not. Note this approach would not
+    // work if we were using nonces instead of hashes.
+    if (ctx.path.endsWith('/')) {
```

Comment:
When manually testing and navigating to something like: `/playground`, I notice a 301 redirect to `/playground/`. Non blocking but curious about the guarantee of a trailing slash?

For example I can still navigate to: https://pr540-ed90bcb---lit-dev-5ftespv5na-uc.a.run.app/playground/index.html

Reply:
Yep we redirect all paths like `/foo` to `/foo/`. But nice catch that you can access the `.html` filenames too! We should block those (https://github.com/lit/lit.dev/blob/48903852d4ae835cf4d9752b0740c7c8fe8c2811/packages/lit-dev-server/src/middleware/redirect-middleware.ts#L17)